substrate/frame/referenda/src/lib.rs

```diff
@@ -239,6 +240,9 @@ pub mod pallet {
 
 		/// The preimage provider.
 		type Preimages: QueryPreimage<H = Self::Hashing> + StorePreimage;
+
+		/// Provider for the block number. Normally this is the `frame_system` pallet.
+		type BlockNumberProvider: BlockNumberProvider<BlockNumber = BlockNumberFor<Self>>;
```

Comment:
Why do we force the block number to be the one in system?
If the block number provider is `RelayChainDataProvider`, then the block number can be different.

It would be more clean to keep it generic here.
Then we should use a new type:
```rust
type BlockNumberFor<T> = <<Self as Config>::BlockNumberProvider as BlockNumberProvider>::BlockNumber;
```

```suggestion
		type BlockNumberProvider: BlockNumberProvider;
```

Reply:
Yeah, we should not require the same block number type.

Reply:
@gui1117 shouldn't it be 

`pub type BlockNumberFor<T, I> = <<T as Config<I>>::BlockNumberProvider as BlockNumberProvider>::BlockNumber;`


since the referenda pallet is an Instantiable one?

Reply:
Yes, we can even name it `ProvidedBlockNumberFor` to avoid conflict with system type alias.